### PR TITLE
[WIP] Improve specs in Access module

### PR DESCRIPTION
**Pull request by eksperimental**
These are a bit more finely-grained specs.
It completes the whole module.
I have tested it with Dialyzer to remove any warnings it could cause.

Docs for Access.get_and_update/3 and callback, have been improved as well.

### CODE REVIEW

File: lib/elixir/lib/access.ex [hunk 1]
```diff
@@ -137,7 +137,8 @@ defmodule Access do
 
   """
 
-  @type t :: list | map | nil | any
+  @type container :: keyword | struct | map | any
+  @type t :: container | nil
```
Comment (eksperimental) [inline lib/elixir/lib/access.ex]: `struct` is to make it more clear that we can have `struct`s in addition to `map`s.
Same as we do with `keyword` and `map` besides having `any`.
But I'm unsure if this is the right approach

Comment (eksperimental) [inline lib/elixir/lib/access.ex]: I was forced to separate it as I couldn't make Dialyzer not complain in `Kernel.pop_in/2`(https://github.com/elixir-lang/elixir/pull/5991/files#diff-e06d5ebc2907d951188e88476f85af51R1974).
It's a bit annoying having the `nil` here as most of the times `nil` will raise so we have to manually create the spec for the function

Comment (josevalim) [inline lib/elixir/lib/access.ex]: Here is how I would do it:

```
@type container ::  keyword | struct | map | any
@type nil_container :: nil
@type t :: container | nil_container
```

File: lib/elixir/lib/access.ex [hunk 2]
```diff
@@ -185,23 +186,26 @@ defmodule Access do
...
-  if not present, and `new_term` is `term` without the key `key`.
+  must be `{value, new_container}` where `value` is the value under `key`
+  (or `nil` if not present) and `new_container` is `container` without the key `key`.
 
   See the implementations of `Map.get_and_update/3` or `Keyword.get_and_update/3`
   for more examples.
   """
-  @callback get_and_update(term :: t, key, (value -> {value, value} | :pop)) :: {value, t}
+  @callback get_and_update(container, key, fun :: (value | nil -> {get_value, update_value} | :pop)) ::
+        {get_value, new_container} when new_container: container, container: t,
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: I don't think `new_container` can be `nil`.

File: lib/elixir/lib/access.ex [hunk 3]
```diff
@@ -215,7 +219,7 @@ defmodule Access do
 
   See the implementations for `Map.pop/3` or `Keyword.pop/3` for more examples.
   """
-  @callback pop(term :: t, key) :: {value, t}
+  @callback pop(container, key) :: {value, container} when container: t
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: Naming the variable `container` and then making it have type `t` is very confusing. We should probably pick a different name. I also don't think the type should be `t` because `nil` is not accepted here.

File: lib/elixir/lib/access.ex [hunk 4]
```diff
@@ -137,7 +137,8 @@ defmodule Access do
 
   """
 
-  @type t :: list | map | nil | any
+  @type container :: keyword | struct | map | any
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: We can probably remove the `any` type, I don't think it is still supported.

File: lib/elixir/lib/access.ex [hunk 5]
```diff
@@ -610,6 +625,8 @@ defmodule Access do
       ** (RuntimeError) Access.at/1 expected a list, got: %{}
 
   """
+  @spec at(non_neg_integer) :: (:get | :get_and_update, list, fun -> {list, list})
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: This type is leaking implementation detail. I don't think most people look at this spec and say "I understand what it is doing". We should probably abstract away the `get | get_and_update` function into a type.

Also note that the first element in `{list, list}` is not required to be a list.

File: lib/elixir/lib/kernel.ex [hunk 6]
```diff
@@ -1971,10 +1971,17 @@ defmodule Kernel do
...
   and the deletion will be considered a success.
   """
-  @spec pop_in(Access.t, nonempty_list(term)) :: {term, Access.t}
+  @spec pop_in(Access.container, nonempty_list(term)) :: {term, Access.container}
   def pop_in(data, keys)
-  def pop_in(nil, [h | _]), do: Access.pop(nil, h)
-  def pop_in(data, keys), do: do_pop_in(data, keys)
+
+  def pop_in(nil, [head | _]) do
+    raise ArgumentError,
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Why this change? Now we have the same logic in two different places. It is ok to call `Access.pop(nil, h)` to error.

File: lib/elixir/lib/access.ex (hunk 3 above)
Comment (eksperimental) [inline lib/elixir/lib/access.ex]: true. the fact is that the `container` type thing came at the very end of my work, as I run Dialyzer and I run into a few issues. I will work on it

File: lib/elixir/lib/access.ex (hunk 4 above)
Comment (eksperimental) [inline lib/elixir/lib/access.ex]: I think it is. at least with tuples.
https://github.com/elixir-lang/elixir/pull/5311

File: lib/elixir/lib/kernel.ex (hunk 6 above)
Comment (eksperimental) [inline lib/elixir/lib/kernel.ex]: the thing is that Dialyzer is complaining because `Access.pop(nil, h)` does not type succeeds.
I guess the best we can do is to create a public function in Kernel that raises. WDYT

File: lib/elixir/lib/access.ex (hunk 3 above)
Comment (eksperimental) [inline lib/elixir/lib/access.ex]: @josevalim what is the reason behind having `nil` as a valid term in `t`?

**Comment by eksperimental** [comment]
@josevalim please have a look at the new changes

File: lib/elixir/lib/kernel.ex (hunk 6 above)
Comment (eksperimental) [inline lib/elixir/lib/kernel.ex]: > lib/kernel.ex:1977:` The call 'Elixir.Access':pop('nil',head@1::any()) will never return since it differs in the 1st argument from the success typing arguments: ([{atom(),_}] | map(),any())

that's the error Dialyzer will give

File: lib/elixir/lib/access.ex (hunk 4 above)
Comment (josevalim) [inline lib/elixir/lib/access.ex]: Yes, i forgot the anonymous function cases. :) Good catch.

File: lib/elixir/lib/access.ex [hunk 7]
```diff
@@ -137,10 +137,24 @@ defmodule Access do
...
+        (:get, structure, (term -> term) ->
+        {get_value, new_structure :: container})
+
+  @type get_and_update_fun(get_value, structure) ::
+        (:get_and_update, structure, (term -> term) ->
+        {get_value, new_structure :: container} | :pop)
+
+  @type access_fun(get_value, structure) ::
+        get_fun(get_value, structure) | get_and_update_fun(get_value, structure)
+
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: ✂️

File: lib/elixir/lib/access.ex [hunk 8]
```diff
@@ -185,37 +199,40 @@ defmodule Access do
...
-  should be `{value, new_term}` where `value` is the value under `key` or `nil`
-  if not present, and `new_term` is `term` without the key `key`.
+  must be `{value, new_structure}` where `value` is the value under `key`
+  (or `nil` if not present) and `new_structure` is `structure` without the key `key`.
 
   See the implementations of `Map.get_and_update/3` or `Keyword.get_and_update/3`
   for more examples.
   """
-  @callback get_and_update(term :: t, key, (value -> {value, value} | :pop)) :: {value, t}
+  @callback get_and_update(structure, key, fun :: (value | nil -> {get_value, update_value} | :pop)) ::
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: Those specs are more precise but they are also much harder to read. There are too many variables. And given we use typespecs mostly for readability purposes, I don't think this is a positive change.

I would propose:

    @callback get_and_update(data, key, (value -> {get_value, update_value} | :pop)) :: {get_value, data} when
              data: container, get_value: value, update_value: value

Don't worry about the nil cases, they were already included when we listed the type as `any`.

File: lib/elixir/lib/access.ex [hunk 9]
```diff
@@ -185,37 +199,40 @@ defmodule Access do
...
+  `key` and `new_structure` is `term` without `key`.
 
-  When the key `key` is not present in the given `term`, a tuple `{value, term}`
+  When the key `key` is not present in the given `structure`, a tuple `{value, structure}`
   should be returned, where `value` is implementation-defined.
 
   See the implementations for `Map.pop/3` or `Keyword.pop/3` for more examples.
   """
-  @callback pop(term :: t, key) :: {value, t}
+  @callback pop(structure, key) :: {value, new_structure :: structure} when structure: container
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: @callback pop(data, key) :: {value, data} when data: container

File: lib/elixir/lib/access.ex [hunk 10]
```diff
@@ -281,21 +300,22 @@ defmodule Access do
...
+  being returned) and the new value to be stored under `key` (`update_value`).
+  `fun` may also return `:pop`, which means the current value
   should be removed from the container and returned.
 
   The returned value is a two-element tuple with the "get" value returned by
   `fun` and a new container with the updated value under `key`.
   """
-  @spec get_and_update(container :: t, key, (value -> {get_value, update_value} | :pop)) ::
-        {get_value, container :: t} when get_value: var, update_value: value
+  @spec get_and_update(container, key, (value | nil -> {get_value, update_value} | :pop)) ::
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: Change it based on the `@callback` please.

File: lib/elixir/lib/access.ex [hunk 11]
```diff
@@ -360,6 +382,11 @@ defmodule Access do
   end
 
   def pop(nil, key) do
+    pop_raise(key)
+  end
+
+  @doc false
+  def pop_raise(key) do
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: If we are going to do this, I would rather `raise` directly on Kernel. Please keep it as it was or raise on kernel.

File: lib/elixir/lib/access.ex [hunk 12]
```diff
@@ -405,6 +432,7 @@ defmodule Access do
       ** (BadMapError) expected a map, got: []
 
   """
+  @spec key(key, term) :: access_fun(term, container) when container: struct | map
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: Please remove the variable.

File: lib/elixir/lib/access.ex [hunk 13]
```diff
@@ -446,6 +474,7 @@ defmodule Access do
       ** (RuntimeError) Access.key!/1 expected a map/struct, got: []
 
   """
+  @spec key!(key) :: access_fun(term, container) when container: struct | map
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: Please remove the variable.

File: lib/elixir/lib/access.ex [hunk 14]
```diff
@@ -137,10 +137,24 @@ defmodule Access do
...
 
+  @type get_fun(get_value, structure) ::
+        (:get, structure, (term -> term) ->
+        {get_value, new_structure :: container})
+
+  @type get_and_update_fun(get_value, structure) ::
+        (:get_and_update, structure, (term -> term) ->
+        {get_value, new_structure :: container} | :pop)
+
+  @type access_fun(get_value, structure) ::
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: The `structure` should be first argument to mimic all other APIs. Also please call it `data` as it is shorter.

File: lib/elixir/lib/access.ex [hunk 15]
```diff
@@ -137,10 +137,24 @@ defmodule Access do
...
+  @type nil_container :: nil
+  @type t :: container | nil_container
   @type key :: any
   @type value :: any
 
+  @type get_fun(get_value, structure) ::
+        (:get, structure, (term -> term) ->
+        {get_value, new_structure :: container})
+
+  @type get_and_update_fun(get_value, structure) ::
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: The `structure` should be first argument to mimic all other APIs. Also please call it `data` as it is shorter.

File: lib/elixir/lib/access.ex [hunk 16]
```diff
@@ -137,10 +137,24 @@ defmodule Access do
...
   """
 
-  @type t :: list | map | nil | any
+  @type container :: keyword | struct | map | any
+  @type nil_container :: nil
+  @type t :: container | nil_container
   @type key :: any
   @type value :: any
 
+  @type get_fun(get_value, structure) ::
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: The `structure` should be first argument to mimic all other APIs. Also please call it `data` as it is shorter.

File: lib/elixir/lib/access.ex [hunk 17]
```diff
@@ -281,21 +300,22 @@ defmodule Access do
...
+  `fun` may also return `:pop`, which means the current value
   should be removed from the container and returned.
 
   The returned value is a two-element tuple with the "get" value returned by
   `fun` and a new container with the updated value under `key`.
   """
-  @spec get_and_update(container :: t, key, (value -> {get_value, update_value} | :pop)) ::
-        {get_value, container :: t} when get_value: var, update_value: value
+  @spec get_and_update(container, key, (value | nil -> {get_value, update_value} | :pop)) ::
+        {get_value, new_container :: container} when container: keyword | struct | map,
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: If we are `container: keyword | struct | map` in multiple places, I would say we need three types:

```
@type container :: keyword | struct | map
@type nil_container :: nil
@type any_container :: any
@type t :: container | any_container | nil_container
```

File: lib/elixir/lib/access.ex [hunk 18]
```diff
@@ -610,6 +641,7 @@ defmodule Access do
       ** (RuntimeError) Access.at/1 expected a list, got: %{}
 
   """
+  @spec at(non_neg_integer) :: access_fun(get_value :: term, container :: list)
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: Much better, thank you!

File: lib/elixir/lib/access.ex (hunk 11 above)
Comment (eksperimental) [inline lib/elixir/lib/access.ex]: The thrid option would be to move the pop_raise/1 function to `Kernel.Utils`.
If that doesn't work, I'm thinking of  leaving the as it was and deal with the Dialyzer error

Comment (josevalim) [inline lib/elixir/lib/access.ex]: Let's go with the previous approach you suggested and copy the error message to Kernel. Thank you!

Comment (eksperimental) [inline lib/elixir/lib/access.ex]: ok. deal then ;-)

**Comment by eksperimental** [comment]
@josevalim this is ready for a whole new review. thank you

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by eksperimental** [comment]
thank you for all the help @josevalim
